lib/saluki-env/src/autodiscovery/providers/local.rs

```diff
@@ -0,0 +1,332 @@
+
+use crate::autodiscovery::{AutodiscoveryEvent, AutodiscoveryProvider, Config};
+
+const BG_MONITOR_INTERVAL: u64 = 30;
+
+/// A local auto-discovery provider that uses the file system.
+pub struct LocalAutoDiscoveryProvider {
+    search_paths: Vec<PathBuf>,
+    sender: Sender<AutodiscoveryEvent>,
+    listener_init: OnceCell<()>,
+}
+
+impl LocalAutoDiscoveryProvider {
+    /// Creates a new `LocalAutoDiscoveryProvider` that will monitor the specified paths.
+    pub fn new<P: AsRef<Path>>(paths: Vec<P>) -> Self {
+        let search_paths: Vec<PathBuf> = paths
+            .iter()
+            .filter_map(|p| {
+                if !p.as_ref().exists() {
+                    warn!("Skipping path '{}' as it does not exist", p.as_ref().display());
+                    return None;
+                }
+                if !p.as_ref().is_dir() {
+                    warn!("Skipping path '{}', it is not a directory.", p.as_ref().display());
+                    return None;
+                }
+                Some(p.as_ref().to_path_buf())
+            })
+            .collect();
+
+        let (sender, _) = broadcast::channel::<AutodiscoveryEvent>(super::AD_STREAM_CAPACITY);
+
+        Self {
+            search_paths,
+            sender,
+            listener_init: OnceCell::new(),
+        }
+    }
+
+    /// Starts a background task that periodically scans for configuration changes
+    async fn start_background_monitor(&self, interval_sec: u64) {
+        let mut interval = interval(Duration::from_secs(interval_sec));
+        let sender = self.sender.clone();
+        let search_paths = self.search_paths.clone();
+
+        info!(
+            "Scanning for local autodiscovery events every {} seconds.",
+            interval_sec
+        );
+
+        tokio::spawn(async move {
+            let mut known_configs = HashSet::new();
+
+            loop {
+                interval.tick().await;
+
+                // Scan for configurations and emit events for changes
+                if let Err(e) = scan_and_emit_events(&search_paths, &mut known_configs, &sender).await {
+                    warn!("Error scanning for configurations: {}", e);
+                }
+            }
+        });
+    }
+}
+
+#[derive(Debug, Deserialize)]
+struct CheckConfig {
+    init_config: serde_yaml::Mapping,
+    instances: Vec<serde_yaml::Mapping>,
+}
+
+/// Parse a YAML file into a Config object
+async fn parse_config_file(path: &PathBuf) -> Result<(String, Config), GenericError> {
+    let content = fs::read_to_string(path).await?;
+
+    let check_config: CheckConfig = match serde_yaml::from_str(&content) {
+        Ok(read) => read,
+        Err(e) => {
+            return Err(GenericError::from(e).context("Failed to decode yaml as check configuration."));
+        }
+    };
+
+    // Build config ID from the file path
+    let config_id = path
+        .canonicalize()?
+        .to_string_lossy()
+        .to_string()
+        .replace(['/', '\\'], "_");
```

Comment:
Right now, this will do synchronous file I/O to canonicalize the path, so this little change just ensures we do it asynchronous (using the `tokio` variants of functions in `tokio::fs`), and additionally, we've skipped the `to_string()` call since `replace` operates on `&str`, so we're allocating the intermediate `String` needlessly since `replace` is going to allocate one anyways.

```suggestion
    let canonicalized_path = fs::canonicalize(&path).await?;
    let config_id = canonicalized_path
        .to_string_lossy()
        .replace(['/', '\\'], "_");
```